Implements Best Match Locator

mahalay opened the pull request:
* Replaces PrioritizedResourceManager
* Adds `suite` option to describe command
* Potentially renders ResourceLocator::getPriority() deprecated

============================================================
Composer allows for classes of the same namespace to be located on multiple paths. Phpspec makes the decision which locator to use based on `priority`. It would be better if the developer is informed that a class being described matches multiple suites and thus be allowed to specify a suite to use. This will make Phpspec easier to work with when used on projects using old frameworks (e.g., Code Igniter).

mheki commented on src/PhpSpec/Locator/BestMatchResourceManager.php:
```diff
@@ -0,0 +1,123 @@
+    {
+        $matchedLocators = [];
+        foreach ($this->locators as $locator) {
+            if (($matchScore = $matchScore = $locator->calculateMatchScore($classname)) > 0) {
```
`$matchScore = $matchScore = ` mistake here

mheki commented on src/PhpSpec/Locator/BestMatchResourceManager.php:
```diff
@@ -0,0 +1,123 @@
+    {
+        $this->locators[] = $locator;
+
+        @usort($this->locators, function (ResourceLocator $locator1, ResourceLocator $locator2) {
```
out of curiosity - why `@` here?

ciaranmcnulty replied:
There's some historical issue that needs it - I think PHP believes the array has changed while being sorted and issues a warning?

mheki commented on spec/PhpSpec/Locator/BestMatchResourceManagerSpec.php:
```diff
@@ -0,0 +1,115 @@
+    }
+
+    function it_locates_all_locators_resources_if_query_string_is_empty($locator1, $locator2,
+                                                                        Resource $resource1, Resource $resource2, Resource $resource3
```
something is messed up with whitespaces here, could you make it more PSR-2 compliant, please?

mahalay commented:
Thanks for the feedback. Corrections were made as suggested.

mheki commented:
thanks @mahalay! Would be great if new features were in sync with the docs. Can you describe your changes in docs, please?

mahalay commented:
Will work on it after office... Why have I forgotten to update the docs. ;)

mahalay commented:
hey @mheki, I've uncovered a scenario where the current implementation will fail. I'll fix that first then finish the docs.

mahalay commented:
Updated the docs and added the scenario where the implementation was failing.

Just curious, what do you guys use for coding standard testing? I tried using squizlabs/php_codesniffer and specified PSR2 as the coding standard and there are a lot of codes with warnings.

ciaranmcnulty commented:
We're using the scrutiniser PRS-2 rules

ciaranmcnulty commented:
@mahalay can you resolve the conflicts please?

ciaranmcnulty commented:
Something now seems to be breaking on PHP 7.2

mahalay commented:
Hey @ciaranmcnulty, installed PHP 7.2 from my end but did not get any errors when I run the tests. Will try to figure out the issue.

ciaranmcnulty commented:
Ah it looks like it's the dev dependencies build that's failing https://travis-ci.org/phpspec/phpspec/jobs/288129896

mahalay commented:
The error was due to the Exception class not being displayed along with the exception message.

ciaranmcnulty commented:
Can you rebase this please?

mahalay commented:
Rebase done, but fails tests on Windows - will setup up a Win VM later
